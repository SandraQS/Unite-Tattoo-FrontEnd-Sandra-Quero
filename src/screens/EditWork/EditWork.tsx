import { useNavigation } from "@react-navigation/core";

import React, { useState } from "react";
import { SafeAreaView, ScrollView, View, TextInput } from "react-native";
import { launchImageLibrary } from "react-native-image-picker";

import { generalStyles } from "../../styles/uniteTatto.styles";
import formsStyles from "../../styles/forms.styles";

import GeneralButton from "../../components/GeneralButton/GeneralButton";
import Title from "../../components/Title/Title";
import NavHeader from "../../components/NavHeader/NavHeader";
import {
  EditWorkScreenNavigationProp,
  EditWorkScreenRouteProp,
} from "../../types/navigation.types";
import { useWorks } from "../../hooks/useWorks";
import RoutesEnum from "../../navigation/routes";
import { INewImage } from "../../types/interfacesComponent";

interface IEditorkProps {
  route: EditWorkScreenRouteProp;
}

export const EditWork = ({ route }: IEditorkProps) => {
  const { editWork } = useWorks();
  const navigation = useNavigation<EditWorkScreenNavigationProp>();

  const {
    params: { work, collection },
  } = route;

  const initialWorkData = {
    ...work,
    tittle: work.tittle,
    tattooArtist: work.tattooArtist,
    description: work.description,
    tattooStyles: work.tattooStyles,
  };
  const [workData, setWorkData] = useState(initialWorkData);
  const [newImage, setNewImage] = useState<INewImage>({
    fileName: "",
    type: "",
    uri: "",
  });

  const textTitle = "EDITAR PROYECTO";

  const onChangeDataWork = (text: string, nameValue: string) => {
    setWorkData({
      ...workData,
      [nameValue]: text,
    });

    return workData;
  };

  const workFormData = new FormData();

  const editClick = () => {
    if (
      newImage.fileName === "" ||
      newImage.type === "" ||
      newImage.uri === ""
    ) {
      editWork(workData, work.id);
    } else {
      workFormData.append("tittle", workData.tittle);
      workFormData.append("tattooArtist", workData.tattooArtist);
      workFormData.append("description", workData.description);
      workFormData.append("tattooStyles", workData.tattooStyles);
      workFormData.append("image", {
        name: newImage.fileName,
        type: newImage.type,
        uri: newImage.uri,
      });
      editWork(workFormData, work.id);
    }

    navigation.navigate(RoutesEnum.works, { collection: collection });
  };

  const chooseFile = () => {
    const options = {
      title: "Select Image",
      customButtons: [
        {
          name: "customOptionKey",
          title: "Choose Photo from Custom Option",
        },
      ],
      storageOptions: {
        skipBackup: true,
        path: "images",
      },
    };

    launchImageLibrary(options, (response: any) => {
      const source = response.assets;
      setNewImage(source[0]);
    });
  };

  return (
    <SafeAreaView style={generalStyles.screenLightBrown}>
      <View style={generalStyles.navHeader}>
        <NavHeader client={false} goBack={() => navigation.pop()} />
      </View>
      <ScrollView>
        <View style={generalStyles.mainContainerGeneral}>
          <View>
            <Title text={textTitle} />
          </View>

          <View style={formsStyles.forms}>
            <View>
              <View>
                <TextInput
                  style={formsStyles.input}
                  value={workData.tittle}
                  placeholder="Titulo"
                  onChangeText={(text) => {
                    onChangeDataWork(text, "tittle");
                  }}
                  accessibilityLabel="tittle"
                  maxLength={20}
                />
              </View>

              <View>
                <TextInput
                  style={formsStyles.input}
                  value={workData.tattooArtist}
                  placeholder="Nombre tatuador"
                  onChangeText={(text) => {
                    onChangeDataWork(text, "tattooArtist");
                  }}
                  accessibilityLabel="tattooArtist"
                  maxLength={12}
                />
              </View>

              <View>
                <TextInput
                  style={[formsStyles.input, formsStyles.textArea]}
                  value={workData.description}
                  placeholder="DescripciÃ³n"
                  onChangeText={(text) => {
                    onChangeDataWork(text, "description");
                  }}
                  numberOfLines={10}
                  multiline={true}
                  accessibilityLabel="description"
                  maxLength={650}
                />
              </View>

              <View>
                <TextInput
                  style={formsStyles.input}
                  value={workData.tattooStyles}
                  placeholder="Estilo"
                  onChangeText={(text) => {
                    onChangeDataWork(text, "tattooStyles");
                  }}
                  accessibilityLabel="tattooStyles"
                  maxLength={20}
                />
              </View>

              <GeneralButton
                textButton="SUBIR IMAGEN"
                functionOnPress={chooseFile}
              />

              <GeneralButton
                textButton="EDITAR PROYECTO"
                functionOnPress={editClick}
              />
            </View>
          </View>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};
