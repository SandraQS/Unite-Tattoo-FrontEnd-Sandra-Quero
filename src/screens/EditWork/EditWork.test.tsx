import { NavigationContainer } from "@react-navigation/native";
import React from "react";
import { Provider } from "react-redux";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { render } from "@testing-library/react-native";
import configureStore from "../../redux/store";
import { EditWork } from "./EditWork";

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
}));

describe("Given EditWork screen", () => {
  describe("When it's rendered", () => {
    test("Then it should show match with the last snapshot", () => {
      const store = configureStore();
      AsyncStorage.getItem = jest.fn().mockResolvedValue({ token: "algo" });
      (global as any).FormData = jest.fn().mockReturnValue({
        append: jest.fn(),
      });
      const route = {
        params: {
          collection: {
            tattooStyles: "Realista",
            image:
              "https://storage.googleapis.com/unite-tattoo.appspot.com/Realista-1638921273323-.png",
            works: [[Object]],
            id: "61aff43a77a75540f78b8449",
          },
          work: {
            tittle: "Lobos - Espalda",
            tattooArtist: "Bruno",
            description:
              "Despue패s de un largo tiempo pude terminar esta espalda, una ma패s para la coleccio패n 游때游녧游낖游녧游낖\n俱데fter quite a long time I was able to finish this back piece, another one!!\n\n郊쀮잺TATTOOING IN @circatattoobcn BARCELONA, SPAIN.\n\n@radiantcolorsink\n@bepantholtattoo\n@tsunami_tattoo_needles",
            tattooStyles: "Realista",
            likes: 0,
            image:
              "https://storage.googleapis.com/unite-tattoo.appspot.com/Realista-1638924401230-.png",
            collectionWork: {
              id: "61aff43a77a75540f78b8449",
            },
          },
        },
      } as any;
       const {
         params: { work, collection },
       } = route;
      new FormData();

      const screen = render(
        <Provider store={store}>
          <NavigationContainer>
            <EditWork route={route} />
          </NavigationContainer>
        </Provider>
      );

      expect(screen).toMatchSnapshot();
    });
  });
});
