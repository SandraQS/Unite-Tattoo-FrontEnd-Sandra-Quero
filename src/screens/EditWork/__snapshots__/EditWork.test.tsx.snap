// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given EditWork screen When it's rendered Then it should show match with the last snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#DFD8D2",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFF",
          "borderBottomColor": "#7C6B53",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 20,
          "paddingTop": 50,
          "paddingVertical": 10,
        }
      }
    >
      <View
        style={Object {}}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            onError={[Function]}
            source={
              Object {
                "testUri": "../../../src/assets/utoo-logo.png",
              }
            }
            style={
              Array [
                Object {
                  "height": 0,
                  "width": 100,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
      <View>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "opacity": 1,
            }
          }
        >
          <View
            style={Object {}}
          >
            <Text
              style={
                Object {
                  "color": "#7C6B53",
                  "fontSize": 15,
                  "paddingEnd": 10,
                }
              }
            >
              Salir
            </Text>
          </View>
          <View
            style={Object {}}
          >
            <Image
              onError={[Function]}
              source={
                Object {
                  "testUri": "../../../src/assets/icono-user.png",
                }
              }
              style={
                Array [
                  Object {
                    "height": 0,
                    "width": 30,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "marginBottom": 100,
            "marginTop": 80,
            "padding": 20,
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
                "marginVertical": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                }
              }
            >
              <Image
                onError={[Function]}
                source={
                  Object {
                    "testUri": "../../../src/assets/icono-rombo-titulo.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "height": 0,
                      "width": 50,
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View
              style={
                Object {
                  "flex": 3,
                  "left": 30,
                  "position": "absolute",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#141414",
                    "fontSize": 20,
                  }
                }
              >
                EDITAR PROYECTO
                <Text
                  style={
                    Object {
                      "fontWeight": "800",
                    }
                  }
                />
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginTop": 30,
            }
          }
        >
          <View>
            <View>
              <TextInput
                accessibilityLabel="tittle"
                maxLength={20}
                onChangeText={[Function]}
                placeholder="Titulo"
                style={
                  Object {
                    "borderBottomColor": "#141414",
                    "borderBottomWidth": 1,
                    "padding": 20,
                  }
                }
                value="Lobos - Espalda"
              />
            </View>
            <View>
              <TextInput
                accessibilityLabel="tattooArtist"
                maxLength={12}
                onChangeText={[Function]}
                placeholder="Nombre tatuador"
                style={
                  Object {
                    "borderBottomColor": "#141414",
                    "borderBottomWidth": 1,
                    "padding": 20,
                  }
                }
                value="Bruno"
              />
            </View>
            <View>
              <TextInput
                accessibilityLabel="description"
                maxLength={650}
                multiline={true}
                numberOfLines={10}
                onChangeText={[Function]}
                placeholder="Descripci칩n"
                style={
                  Array [
                    Object {
                      "borderBottomColor": "#141414",
                      "borderBottomWidth": 1,
                      "padding": 20,
                    },
                    Object {
                      "paddingTop": 20,
                    },
                  ]
                }
                value="Despue패s de un largo tiempo pude terminar esta espalda, una ma패s para la coleccio패n 游때游녧游낖游녧游낖
俱데fter quite a long time I was able to finish this back piece, another one!!

郊쀮잺TATTOOING IN @circatattoobcn BARCELONA, SPAIN.

@radiantcolorsink
@bepantholtattoo
@tsunami_tattoo_needles"
              />
            </View>
            <View>
              <TextInput
                accessibilityLabel="tattooStyles"
                maxLength={20}
                onChangeText={[Function]}
                placeholder="Estilo"
                style={
                  Object {
                    "borderBottomColor": "#141414",
                    "borderBottomWidth": 1,
                    "padding": 20,
                  }
                }
                value="Realista"
              />
            </View>
            <View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "borderColor": "black",
                    "borderWidth": 1,
                    "marginBottom": 20,
                    "marginTop": 20,
                    "opacity": 1,
                    "padding": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "black",
                      "fontSize": 18,
                      "fontWeight": "900",
                      "textAlign": "center",
                    }
                  }
                >
                  SUBIR IMAGEN
                </Text>
              </View>
            </View>
            <View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "borderColor": "black",
                    "borderWidth": 1,
                    "marginBottom": 20,
                    "marginTop": 20,
                    "opacity": 1,
                    "padding": 20,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "black",
                      "fontSize": 18,
                      "fontWeight": "900",
                      "textAlign": "center",
                    }
                  }
                >
                  EDITAR PROYECTO
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
